src/auth/defaultawssecuritycredentialssupplier.ts

```diff
@@ -0,0 +1,265 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {ExternalAccountSupplierContext} from './baseexternalclient';
+import {Gaxios, GaxiosOptions} from 'gaxios';
+import {Transporter} from '../transporters';
+import {AwsSecurityCredentialsSupplier} from './awsclient';
+import {AwsSecurityCredentials} from './awsrequestsigner';
+import {Headers} from './oauth2client';
+
+/**
+ * Interface defining the AWS security-credentials endpoint response.
+ */
+interface AwsSecurityCredentialsResponse {
+  Code: string;
+  LastUpdated: string;
+  Type: string;
+  AccessKeyId: string;
+  SecretAccessKey: string;
+  Token: string;
+  Expiration: string;
+}
+
+/**
+ * Internal AWS security credentials supplier implementation used by {@link AwsClient}
+ * when a credential source is provided instead of a user defined supplier.
+ * The logic is summarized as:
+ * 1. If imdsv2_session_token_url is provided in the credential source, then
+ *    fetch the aws session token and include it in the headers of the
+ *    metadata requests. This is a requirement for IDMSv2 but optional
+ *    for IDMSv1.
+ * 2. Retrieve AWS region from availability-zone.
+ * 3a. Check AWS credentials in environment variables. If not found, get
+ *     from security-credentials endpoint.
+ * 3b. Get AWS credentials from security-credentials endpoint. In order
+ *     to retrieve this, the AWS role needs to be determined by calling
+ *     security-credentials endpoint without any argument. Then the
+ *     credentials can be retrieved via: security-credentials/role_name
+ * 4. Generate the signed request to AWS STS GetCallerIdentity action.
+ * 5. Inject x-goog-cloud-target-resource into header and serialize the
+ *    signed request. This will be the subject-token to pass to GCP STS.
+ */
+export class DefaultAwsSecurityCredentialsSupplier
+  implements AwsSecurityCredentialsSupplier
+{
+  private readonly regionUrl?: string;
+  private readonly securityCredentialsUrl?: string;
+  private readonly imdsV2SessionTokenUrl?: string;
+
+  /**
+   * Instantiates a new DefaultAwsSecurityCredentialsSupplier using information
+   * from the credential_source stored in the ADC file.
+   * @param regionUrl The URL to call to retrieve the active AWS region.
+   * @param securityCredentialsUrl The URL to call to retrieve AWS security credentials.
+   * @param imdsV2SessionTokenUrl The URL to call to retrieve the IMDSV2 session token.
+   */
+  constructor(
+    regionUrl?: string,
+    securityCredentialsUrl?: string,
+    imdsV2SessionTokenUrl?: string
+  ) {
+    this.regionUrl = regionUrl;
+    this.securityCredentialsUrl = securityCredentialsUrl;
+    this.imdsV2SessionTokenUrl = imdsV2SessionTokenUrl;
+  }
```

Comment:
Will change to use an options object, for the transporter - my thought was that it would be good to give users who are going to be implementing the suppler interface the transporter when called rather than forcing them to construct the suppliers with it since they may be instantiating the supplier before the client and its transporter are set up, and was just following that pattern with the internal suppliers. Thoughts?